keps/sig-node/4742-node-topology-downward-api/README.md

```diff
@@ -0,0 +1,409 @@
+              fieldPath: metadata.labels['topology.kubernetes.io/zone']
+```
+
+### User Stories
+
+* As an ML engineer, I want to optimize GPU-to-GPU communication during training which requires topology-awareness in my training code.
+* As a database developer, I want to leverage topology information to improve fault tolerance of sharded databases on Kubernetes.
+* As a developer of a Kubernetes CNI plugin, I want to pass topology information down to the CNI plugin to optimize data paths for container networks.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+* Scope creep. Allowing additional node information or node label info could
+create security issues. This is mitigated by limiting the node labels
+to strictly those that are standardized through KEP-1659.
+
+* Exposing sensitive data as node labels to pods. This is mitigated by ensuring
+standard topology labels are available to Pods.
+
+* Stale data. Information obtained through node labels is like information
+attained through a configmap or secret mounted to a pod, being passed on
+creation but not guaranteed to be immutable and thus should be treated as so.
+
+
+## Design Details
+
+* A built-in Kubernetes admission plugin, `PodTopologyLabels` will be introduced in kube-apiserver
+* The `PodTopologyLabels` admission plugin is responsible for mutating `pods/binding` subresource, adding topology labels matching the target Node.
+* `PodTopologyLabels` admission will not override any existing topology labels on Pods.
+* A feature gate, `PodTopologyLabelsAdmission` will be introduced in v1.33. Alpha and disabled by default.
+The `PodTopologyLabels` admission plugin can only be set when this feature gate is enabled.
+* The Binding REST implementation will be updated to copy all labels from `pods/binding` subresource into Pods.
+At this point we do not remove labels, it is assumed that any labels in Binding are safe to expose via Downward API.
+* For exposing additional node labels, at the discretion of the cluster admin, a mutating admission webhook can be used to mutate labels of the `pods/binding` subresource.
+
+### Test Plan
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+Unit tests will be added for:
+* New `PodTopologyLabels` admission plugin
+* Binding REST implementation
+
+Unit tests will also ensure behavior is exercised when the feature gate is enabled.
+
+##### Integration tests
+
+Integration tests will be added to test the following behavior:
+* Pods contain topology labels when `PodTopologyLabels` admission plugin is enabled.
+* Topology labels can be expressed in Pod Downward API
+* Node labels outside standard topology labels are disallowed
+* Topology labels are empty if underlying Node does not specify a topology
```

Comment:
can we add a test that ensures that customer can use their own webhook to copy or set those values

Reply:
Not sure I'd be willing to write an e2e test that involves introducing an external component. If we can get it working with something like MutatingAdmissionPolicy then yes.

Reply:
Added an e2e test for using MutatingAdmissionPolicy to test custom node labels